### PR TITLE
On Wayland, pre-multiply alpha for custom cursor

### PR DESCRIPTION
**Pull request by kchibisov**
Fixes: #3360

- [x] Tested on all platforms changed
- [x] Added an entry to `CHANGELOG.md` if knowledge of this change could be valuable to users
- [ ] Updated documentation to reflect any user-facing changes, including notes of platform-specific behavior
- [ ] Created or updated an example program if it would help users understand this functionality
- [ ] Updated [feature matrix](https://github.com/rust-windowing/winit/blob/master/FEATURES.md), if new features were added or implemented

--

@daxpedda  I guess we should document whether alpha is premultiplied or not? The assumption is that it's probably not premultiplied?

### CODE REVIEW

**Comment by daxpedda** [comment]
Yes, it should not be pre-multiplied and you are correct, we should most definitely be documented.

Web already presumes it's not pre-multiplied:
https://github.com/rust-windowing/winit/blob/df8805c0d27ca3172f4fa1a6bc1bf9709c4c8d0b/src/platform_impl/web/cursor.rs#L602

**Comment by kchibisov** [comment]
I guess that should work?

File: src/platform/web.rs [hunk 1]
```diff
@@ -248,6 +248,8 @@ pub trait CustomCursorExtWebSys {
     /// It uses the [url css function](https://developer.mozilla.org/en-US/docs/Web/CSS/url),
     /// but browser support for image formats is inconsistent. Using [PNG] is recommended.
     ///
+    /// The alpha channel is assumed to be **not** premultiplied.
+    ///
```
Comment (daxpedda) [inline src/platform/web.rs:252]:
```suggestion
```
Not relevant here, because we don't parse the image, we pass it straight on as a URL to CSS.

**Review by daxpedda — CHANGES_REQUESTED** [review]
Otherwise LGTM.

File: src/platform/web.rs (hunk 1 above)
Comment (kchibisov) [reply src/platform/web.rs:252]: Hm, but you set that alpha is not premultiplied or you just pass as is when it's loaded like that?

Comment (daxpedda) [reply src/platform/web.rs:252]: When it's loaded from a URL it bypasses that part of the code indeed.

**Review by daxpedda — APPROVED** [review]
Doc changes approved!

File: src/platform_impl/linux/wayland/types/cursor.rs [hunk 2]
```diff
@@ -37,12 +37,15 @@ impl CustomCursor {
...
+        for (canvas_chunk, rgba) in canvas.chunks_exact_mut(4).zip(image.rgba.chunks_exact(4)) {
+            // Alpha in buffer is premultiplied.
+            let alpha = rgba[3] as f32 / 255.;
+            let r = (rgba[0] as f32 * alpha) as u32;
+            let g = (rgba[1] as f32 * alpha) as u32;
+            let b = (rgba[2] as f32 * alpha) as u32;
+            let color = ((rgba[3] as u32) << 24) + (r << 16) + (g << 8) + b;
+            let array: &mut [u8; 4] = canvas_chunk.try_into().unwrap();
+            *array = color.to_le_bytes();
         }
```
Comment (notgull) [inline src/platform_impl/linux/wayland/types/cursor.rs:49]: I would add a comment here that this should be SIMD-optimized once the standard library exposes the proper primitives.

**Review by notgull — APPROVED** [review]

File: src/platform_impl/linux/wayland/types/cursor.rs (hunk 2 above)
Comment (kchibisov) [reply src/platform_impl/linux/wayland/types/cursor.rs:49]: compilers are good in figuring out such loops, just pointless complexity.

**Pull request merged**